### ISSUE TITLE
False-positive: "Objects are always passed by reference" with interface & indirection

### ISSUE THREAD

**Issue by guilliamxavier** [kalessil/phpinspectionsea#1340]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Plugin         | Php Inspections (EA Extended), version v3.0.13 |
| Language level | PHP 7.2                        |

Current behaviour
--

Code (which runs just fine):
```php
<?php
interface I {
    public function out(DateTime &$out = null): bool;
    public function in(DateTime $in): void;
}
class C implements I {
    protected function doOut(DateTime &$out = null): bool {
        $out = new DateTime();
        return true;
    }
    public function out(DateTime &$out = null): bool {
        return $this->doOut($out);
    }
    public function in(DateTime $in): void {
        echo $in->format('c'), PHP_EOL;
    }
}
final class Main {
    public static function run(I $i): void {
        if ($i->out($d)) {
            $i->in($d);
        }
    }
}
Main::run(new C());
```
triggers inspection warning "Objects are always passed by reference; please correct '& $out'." on both instances of `function out(DateTime &$out = null)` (and suggests a quickfix "Cleanup parameter definition" that changes to `DateTime $out = null` which obviously is wrong) (but not on `function doOut`, I guess thanks to the assignment in the body).

Expected behaviour
--
No inspection warning.

Environment details
---

PhpStorm 2019.1.2
Build #PS-191.7141.52, built on May 8, 2019

**Comment by kalessil** [comment]
Confirmed, will take care of fixing the bug.

**Comment by guilliamxavier** [comment]
Thanks for your feedback, but did you intend to close?

BTW, the two cases are actually independent:

1. interface:
   
   ```php
   interface I {
       public function out(DateTime &$out = null): bool;
   }
   ```
   
   (because we cannot assume the implementation)

2. indirection:
   
   ```php
   class C {
       protected function doOut(DateTime &$out = null): bool {
           $out = new DateTime();
           return true;
       }
       public function out(DateTime &$out = null): bool {
           return $this->doOut($out);
       }
   }
   ```
   
   (because `out` passes `$out` to `doOut` which does takes it by reference).
   
   Also with this chain:
   
   ```php
   function f(DateTime &$d): bool {
       return g($d);
   }
   function g(DateTime &$d): bool {
       return h($d);
   }
   function h(DateTime &$d): bool {
       $d->modify('+1 day');
       return true;
   }
   ```
   
   I would expect a warning only on `function h`

**Comment by guilliamxavier** [comment]
@kalessil Sorry to bring this (low-priority) back again, but it does look like this was closed by mistake, and I cannot reopen; can you?

**Comment by kalessil** [comment]
Re-opened

**Comment by kalessil** [comment]
Fixed!

**Comment by guilliamxavier** [comment]
That was fast, thanks!
I see that the actual fix (checking for a default value) is simpler than what I imagined, but the end result is the same: I confirm that the original issue is fixed in v3.0.16 🙂
